[inductor] Support using the 'stream' param in AOT mode

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105589

Summary:
When in AOT mode, make use of the existing stream param:
- Pass through and use the stream param in the launchKernel helper function.
- In non-AOT mode, assign the stream param in the caller and pass to launchKernel
- Use a CUDAStreamGuard so all fallback ops execute on the stream
- CUDAStreamGuard subsumes CUDAGuard in AOT mode since it sets both stream and device

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) In AOT mode:
```
   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes, stream, args, nullptr));

   ...
   at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream);
   ...
```

   b) Regular cpp wrapper:
```
   ...
   at::cuda::CUDAGuard device_guard(0);
   cudaStream_t stream0 = at::cuda::getCurrentCUDAStream(0);
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream0);
   ...
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

masnesral committed:
[inductor] Support using the 'stream' param in AOT mode

Summary:
When in AOT mode, make use of the existing stream param:
- Use a CUDAStreamGuard so all fallback ops execute on the stream
- Pass and use the stream param in the launchKernel helper function

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) The launchKernel code in AOT mode:

   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           void* args[],
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes,
           stream, args, nullptr));

   ...
   launchKernel(triton_poi_fused_convolution_max_pool2d_with_indices_relu_3, 365, 1, 1, 8, 0, kernel_args_var_3, stream);
   ...

   b) Using the stream guard:

       at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));

[ghstack-poisoned]

masnesral commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1312,11 +1328,11 @@ def write_header(self):
                     int sharedMemBytes,
                     void* args[],
-                    int device_index) {
+                    {stream_param}) {{
```
@desertfire Do you mean that you'd like the parameters for the launchKernel helper to stay the same between the two modes, or do you mean you'd like the call to cuLaunchKernel to look the same between the two modes (or both of those things)?

desertfire replied:
I mean  the parameter declaration for the launchKernel helper to stay the same between the two modes.

masnesral replied:
Well, we only have the `stream` param available in the calling function when in aot mode. Do you want to do something like this:

   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           void* args[],
           int device_index,
-->     cudaStream_t stream = nullptr
           ) {
       auto stream = at::cuda::getCurrentCUDAStream(device_index);
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes, stream, args, nullptr));
   }

masnesral replied:
Oh, perhaps you mean to create the stream in the caller?

desertfire replied:
> Oh, perhaps you mean to create the stream in the caller?

correct

masnesral committed:
Update on "[inductor] Support using the 'stream' param in AOT mode"


Summary:
When in AOT mode, make use of the existing stream param:
- Use a CUDAStreamGuard so all fallback ops execute on the stream
- Pass and use the stream param in the launchKernel helper function

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) The launchKernel code in AOT mode:

   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           void* args[],
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes,
           stream, args, nullptr));

   ...
   launchKernel(triton_poi_fused_convolution_max_pool2d_with_indices_relu_3, 365, 1, 1, 8, 0, kernel_args_var_3, stream);
   ...

   b) Using the stream guard:

       at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented:
To fix lint errors, you can run `make setup_lint` to install the linter, and then `lintrunner -a`.

masnesral committed:
Update on "[inductor] Support using the 'stream' param in AOT mode"


Summary:
When in AOT mode, make use of the existing stream param:
- Use a CUDAStreamGuard so all fallback ops execute on the stream
- Pass and use the stream param in the launchKernel helper function

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) The launchKernel code in AOT mode:

   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           void* args[],
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes,
           stream, args, nullptr));

   ...
   launchKernel(triton_poi_fused_convolution_max_pool2d_with_indices_relu_3, 365, 1, 1, 8, 0, kernel_args_var_3, stream);
   ...

   b) Using the stream guard:

       at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support using the 'stream' param in AOT mode"


Summary:
When in AOT mode, make use of the existing stream param:
- Use a CUDAStreamGuard so all fallback ops execute on the stream
- Pass and use the stream param in the launchKernel helper function

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) The launchKernel code in AOT mode:

   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           void* args[],
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes,
           stream, args, nullptr));

   ...
   launchKernel(triton_poi_fused_convolution_max_pool2d_with_indices_relu_3, 365, 1, 1, 8, 0, kernel_args_var_3, stream);
   ...

   b) Using the stream guard:

       at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support using the 'stream' param in AOT mode"


Summary:
When in AOT mode, make use of the existing stream param:
- Pass through and use the stream param in the launchKernel helper function.
-- In non-AOT mode, /assign the stream param in the caller and pass to launchKernel
- Use a CUDAStreamGuard so all fallback ops execute on the stream
-- This usage subsumes CUDAGuard in AOT mode since it sets both stream and device

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) In AOT mode:
```
   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes, stream, args, nullptr));

   ...
   at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream);
   ...
```

   b) Regular cpp wrapper:
```
   cudaStream_t stream;
   ...
   at::cuda::CUDAGuard device_guard(0);
   stream = at::cuda::getCurrentCUDAStream(0);
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream);
   ...
```

[ghstack-poisoned]

masnesral committed:
fix lints on "[inductor] Support using the 'stream' param in AOT mode"


Summary:
When in AOT mode, make use of the existing stream param:
- Pass through and use the stream param in the launchKernel helper function.
-- In non-AOT mode, /assign the stream param in the caller and pass to launchKernel
- Use a CUDAStreamGuard so all fallback ops execute on the stream
-- This usage subsumes CUDAGuard in AOT mode since it sets both stream and device

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) In AOT mode:
```
   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes, stream, args, nullptr));

   ...
   at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream);
   ...
```

   b) Regular cpp wrapper:
```
   cudaStream_t stream;
   ...
   at::cuda::CUDAGuard device_guard(0);
   stream = at::cuda::getCurrentCUDAStream(0);
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream);
   ...
```

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -984,6 +1000,12 @@ def write_wrapper_decl(self):
+            if not V.graph.aot_mode:
+                self.wrapper_call.splice(
+                    """
+                    cudaStream_t stream;
```
This should go to CudaWrapperCodeGen. CppWrapperCodeGen is shared for cpu and gpu. I know we are already mixing the logic at some places, but let's not add more to that.

masnesral replied:
@desertfire , sure. I put it here to keep consistency because we write the stream param in the parameter list at the top of this function (line 950). So by the same logic, shouldn't that go in CudaWrapperCodeGen as well? If so, do you want me to refactor this write_wrapper_decl method and specialize in CudaWrapperCodeGen vs. CppWrapperCodeGen?

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -147,10 +147,26 @@ class EnterCudaDeviceContextManagerLine:
+                else:
+                    code.writeline(f"device_guard.set_index({self.device_idx});")
+                code.writeline(
+                    f"stream = at::cuda::getCurrentCUDAStream({self.device_idx});"
```
You can follow how stream is get and passed in the python wrapper codegen, https://github.com/pytorch/pytorch/blob/a361fceef39766e839fbfe813d195339c66b5391/torch/_inductor/codegen/wrapper.py#L717C1-L717C1

masnesral committed:
addressed @desertfire's comments; fixed lints; re-ran failing tests on "[inductor] Support using the 'stream' param in AOT mode"


Summary:
When in AOT mode, make use of the existing stream param:
- Pass through and use the stream param in the launchKernel helper function.
- In non-AOT mode, assign the stream param in the caller and pass to launchKernel
- Use a CUDAStreamGuard so all fallback ops execute on the stream
- CUDAStreamGuard subsumes CUDAGuard in AOT mode since it sets both stream and device

Test Plan:
- Ran cpp_wrapper tests: pytest test/inductor/test_cpp_wrapper.py
- Manually inspected cpp output from the alexnet benchmark:

  a) In AOT mode:
```
   static inline void launchKernel(
           CUfunction func,
           int gridX,
           int gridY,
           int gridZ,
           int numWraps,
           int sharedMemBytes,
           cudaStream_t stream) {
       AT_CUDA_DRIVER_CHECK_OVERRIDE(cuLaunchKernel(
           func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes, stream, args, nullptr));

   ...
   at::cuda::CUDAStreamGuard stream_guard(at::cuda::getStreamFromExternal(stream, 0));
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream);
   ...
```

   b) Regular cpp wrapper:
```
   ...
   at::cuda::CUDAGuard device_guard(0);
   cudaStream_t stream0 = at::cuda::getCurrentCUDAStream(0);
   ...
   launchKernel(triton_poi_fused_convolution_0, 1, 784, 1, 4, 4352, kernel_args_var_0, stream0);
   ...
```

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
